# Data Quality & Cleaning

This notebook focuses on validating and improving the quality of the FinSight source data before further analysis and modeling.

## Objectives

- identify duplicate records and validate key uniqueness,
- analyze missing values,
- correct inconsistent data types,
- validate relationships between datasets,
- detect invalid or anomalous values,
- build reusable data validation checks,
- generate a data quality summary,
- save cleaned datasets to `data/processed/`.

> Raw data in `data/raw/` remains unchanged throughout the process.

In [57]:
import pandas as pd
import numpy as np
import json

print("FinSight environment ready")

FinSight environment ready


## Load Source Data

In [58]:
users = pd.read_csv("../data/raw/users_data.csv")
transactions = pd.read_csv("../data/raw/transactions_data.csv")
cards = pd.read_csv("../data/raw/cards_data.csv")
with open("../data/raw/mcc_codes.json", "r") as file:
    mcc_dict = json.load(file)
codes = pd.DataFrame(
    mcc_dict.items(),
    columns=["mcc", "description"]
)
with open("../data/raw/train_fraud_labels.json", "r") as file:
    fraud_dict = json.load(file)
fraud_target = fraud_dict["target"]
fraud = pd.DataFrame(
    fraud_target.items(),
    columns=["transaction_id", "is_fraud"]
)

print("Loading raw files ready")

Loading raw files ready


## Duplicate Validation

### Users

In [59]:
users.duplicated().sum()

np.int64(0)

**Result:** No fully duplicated rows were found in `users`.

In [60]:
users["id"].duplicated().sum()

np.int64(0)

**Result:** No duplicate `id` values were found. Each user has a unique identifier.

### Cards

In [61]:
cards.duplicated().sum()

np.int64(0)

**Result:** No fully duplicated rows were found in `cards`.

#### Primary Key Duplicate Check

Check whether `id` contains duplicate values, as each card should have a unique identifier.

In [62]:
cards["id"].duplicated().sum()

np.int64(0)

**Result:** No duplicate `id` values were found. Each card has a unique identifier.

### Transactions

In [63]:
transactions.duplicated().sum()

np.int64(0)

**Result:** No fully duplicated rows were found in `transactions`.

In [64]:
transactions["id"].duplicated().sum()

np.int64(0)

**Result:** No duplicate `id` values were found. Each transaction has a unique identifier.

#### MCC Codes

In [65]:
codes.duplicated().sum()

np.int64(0)

**Result:** No fully duplicated rows were found in `codes`.

In [66]:
codes["mcc"].duplicated().sum()

np.int64(0)

**Result:** No duplicate `mcc` values were found. Each code has a unique identifier.

#### Fraud Labels

In [67]:
fraud.duplicated().sum()

np.int64(0)

**Result:** No fully duplicated rows were found in `fraud`.

In [68]:
fraud["transaction_id"].duplicated().sum()

np.int64(0)

**Result:** No duplicate `transaction_id` values were found. Each fraud has a unique identifier.

## Missing Values Validation

#### Users

In [69]:
users.isna().sum()

id                   0
current_age          0
retirement_age       0
birth_year           0
birth_month          0
gender               0
address              0
latitude             0
longitude            0
per_capita_income    0
yearly_income        0
total_debt           0
credit_score         0
num_credit_cards     0
dtype: int64

**Result:** No missing values were found in `users`.

#### Cards

In [70]:
cards.isna().sum()

id                       0
client_id                0
card_brand               0
card_type                0
card_number              0
expires                  0
cvv                      0
has_chip                 0
num_cards_issued         0
credit_limit             0
acct_open_date           0
year_pin_last_changed    0
card_on_dark_web         0
dtype: int64

**Result:** No missing values were found in `cards`.

#### Transactions

In [71]:
transactions.isna().sum()

id                       0
date                     0
client_id                0
card_id                  0
amount                   0
use_chip                 0
merchant_id              0
merchant_city            0
merchant_state     1563700
zip                1652706
mcc                      0
errors            13094522
dtype: int64

#### Missing Values Percentage

Calculate the percentage of missing values in each column of `transactions`.

In [72]:
percentage = round(((transactions.isna().sum())/len(transactions)) * 100, 2)
percentage

id                 0.00
date               0.00
client_id          0.00
card_id            0.00
amount             0.00
use_chip           0.00
merchant_id        0.00
merchant_city      0.00
merchant_state    11.75
zip               12.42
mcc                0.00
errors            98.41
dtype: float64

#### Transaction Errors

Inspect the values present in `errors` to determine whether missing values represent transactions without an error.

In [73]:
transactions["errors"].value_counts()

errors
Insufficient Balance                                   130902
Bad PIN                                                 32119
Technical Glitch                                        26271
Bad Card Number                                          7767
Bad Expiration                                           6161
Bad CVV                                                  6106
Bad Zipcode                                              1126
Bad PIN,Insufficient Balance                              293
Insufficient Balance,Technical Glitch                     243
Bad Card Number,Insufficient Balance                       71
Bad PIN,Technical Glitch                                   70
Bad CVV,Insufficient Balance                               57
Bad Expiration,Insufficient Balance                        47
Bad Card Number,Bad CVV                                    38
Bad Card Number,Bad Expiration                             33
Bad Expiration,Bad CVV                                     32
B

#### Transaction Errors — Missing Values

Check whether missing values in `errors` represent transactions with no recorded error by including missing values in the frequency count.

In [74]:
transactions["errors"].value_counts(dropna = False)

errors
NaN                                                    13094522
Insufficient Balance                                     130902
Bad PIN                                                   32119
Technical Glitch                                          26271
Bad Card Number                                            7767
Bad Expiration                                             6161
Bad CVV                                                    6106
Bad Zipcode                                                1126
Bad PIN,Insufficient Balance                                293
Insufficient Balance,Technical Glitch                       243
Bad Card Number,Insufficient Balance                         71
Bad PIN,Technical Glitch                                     70
Bad CVV,Insufficient Balance                                 57
Bad Expiration,Insufficient Balance                          47
Bad Card Number,Bad CVV                                      38
Bad Card Number,Bad Expiration   

**Result:** The `errors` column contains 98.41% missing values. These values represent transactions with no recorded error rather than a data quality issue. Non-null values contain specific transaction error types.

#### Missing `merchant_state` Investigation

Inspect transactions with missing `merchant_state` values to determine whether the missingness is expected or indicates a data quality issue.

In [75]:
transactions[transactions["merchant_state"].isna()].head(5)

,id,date,client_id,card_id,amount,use_chip,merchant_id,merchant_city,merchant_state,zip,mcc,errors
7,7475335,2010-01-01 00:14:00,1684,2140,$26.46,Online Transaction,39021,ONLINE,NaN,NaN,4784,NaN
8,7475336,2010-01-01 00:21:00,335,5131,$261.58,Online Transaction,50292,ONLINE,NaN,NaN,7801,NaN
18,7475346,2010-01-01 00:34:00,394,4717,$26.04,Online Transaction,39021,ONLINE,NaN,NaN,4784,NaN
24,7475353,2010-01-01 00:43:00,301,3742,$10.17,Online Transaction,39021,ONLINE,NaN,NaN,4784,NaN
26,7475356,2010-01-01 00:45:00,566,3439,$16.86,Online Transaction,16798,ONLINE,NaN,NaN,4121,NaN


#### Missing `merchant_state` by Transaction Type

Check whether missing `merchant_state` values occur exclusively in online transactions.

In [76]:
transactions[
    transactions["merchant_state"].isna()
]["use_chip"].unique()

<StringArray>
['Online Transaction', 'Chip Transaction']
Length: 2, dtype: str

#### Chip Transactions with Missing `merchant_state`

Inspect chip transactions where `merchant_state` is missing to determine whether there is another pattern explaining the missing location data.

In [77]:
transactions[
    (transactions["merchant_state"].isna()) &
    (transactions["use_chip"] == "Chip Transaction")
].head()

,id,date,client_id,card_id,amount,use_chip,merchant_id,merchant_city,merchant_state,zip,mcc,errors
6573529,15473515,2015-01-01 12:19:00,944,2828,$-134.00,Chip Transaction,13348,ONLINE,NaN,NaN,4722,NaN
6573565,15473561,2015-01-01 12:27:00,944,2828,$358.94,Chip Transaction,13348,ONLINE,NaN,NaN,4722,NaN
6573594,15473593,2015-01-01 12:32:00,1117,1162,$117.43,Chip Transaction,52073,ONLINE,NaN,NaN,4722,NaN
6573677,15473695,2015-01-01 12:51:00,1117,1162,$-147.00,Chip Transaction,52073,ONLINE,NaN,NaN,4722,NaN
6574120,15474225,2015-01-01 14:24:00,972,6052,$233.77,Chip Transaction,13348,ONLINE,NaN,NaN,4722,NaN


#### Missing `merchant_state` by Merchant Location

Check whether transactions with missing `merchant_state` are associated with `merchant_city = ONLINE`.

In [78]:
transactions[
    transactions["merchant_state"].isna()
]["merchant_city"].value_counts()

merchant_city
ONLINE    1563700
Name: count, dtype: int64

**Result:** All 1,563,700 transactions with missing `merchant_state` have `merchant_city = ONLINE`. Therefore, the missing values are expected for online transactions and do not indicate a data quality issue.

#### Missing `zip` Investigation

Inspect transactions with missing `zip` values to determine whether the missingness is expected or indicates a data quality issue.

In [79]:
transactions[
    transactions["zip"].isna()
].head()

,id,date,client_id,card_id,amount,use_chip,merchant_id,merchant_city,merchant_state,zip,mcc,errors
7,7475335,2010-01-01 00:14:00,1684,2140,$26.46,Online Transaction,39021,ONLINE,NaN,NaN,4784,NaN
8,7475336,2010-01-01 00:21:00,335,5131,$261.58,Online Transaction,50292,ONLINE,NaN,NaN,7801,NaN
18,7475346,2010-01-01 00:34:00,394,4717,$26.04,Online Transaction,39021,ONLINE,NaN,NaN,4784,NaN
24,7475353,2010-01-01 00:43:00,301,3742,$10.17,Online Transaction,39021,ONLINE,NaN,NaN,4784,NaN
26,7475356,2010-01-01 00:45:00,566,3439,$16.86,Online Transaction,16798,ONLINE,NaN,NaN,4121,NaN


#### Missing `zip` by Merchant Location

Check which merchant locations are associated with missing `zip` values.

In [80]:
transactions[
    transactions["zip"].isna()
]["merchant_city"].value_counts()

merchant_city
ONLINE            1563700
Cancun               9325
Rome                 7081
Mexico City          5102
Cabo San Lucas       4550
                   ...   
Addis Ababa             3
Libreville              3
Paramaribo              3
Nuku Alofa              2
Baku                    2
Name: count, Length: 169, dtype: int64

#### Missing `zip` by Merchant State

Inspect `merchant_state` values for transactions with missing `zip` to determine whether the additional missing ZIP codes are associated with international transactions.

In [81]:
transactions[
    transactions["zip"].isna()
]["merchant_state"].value_counts(dropna=False)

merchant_state
NaN               1563700
Mexico              27401
Canada              10647
Italy                7081
United Kingdom       4482
                   ...   
Ethiopia                3
Gabon                   3
Suriname                3
Tonga                   2
Azerbaijan              2
Name: count, Length: 148, dtype: int64

**Result:** Missing `zip` values are primarily associated with online and international transactions. Online transactions have no physical merchant location, while international transactions may not use the ZIP code format represented by this field. Therefore, these missing values appear to be structurally expected rather than a data quality issue.

#### Mcc codes

In [85]:
codes.isna().sum()

mcc            0
description    0
dtype: int64

**Result:** No missing values were found in `codes`.

#### Fraud Labels

In [87]:
fraud.isna().sum()

transaction_id    0
is_fraud          0
dtype: int64

In [ ]:
**Result:** No missing values were found in `fraud`.